### PR TITLE
feat: data encryption support more plugins

### LINKED ISSUE #8407: Proposal: APISIX Supports Global Data Encryption
# Background
Nowadays, if a plugin needs to store sensitive information, it will design a separate set of encryption mechanism, which causes the problem of repeated development and cumbersome management. Now we consider introducing Data Encryption function, hereinafter referred to as DE, to encrypt the specified information, and any module/plugin that needs to encrypt sensitive information only needs to specify the fields to be encrypted on display, so that transparent encryption/decryption can be performed when saving/reading the information.

# Benefits
Unify the encryption mechanism of APISIX to avoid duplication of development while improving product competitiveness

# Goals 

APISIX provides DE functionality to support the use of this capability in plugins and elsewhere to protect sensitive information.

## The user scenarios are as follows

1. Turn on the DE function in the configuration file

```YAML
apisix:
    enable_global_data_encryption: true
```

2. set encrypted fields

```Ada
local schema = {
    type = "object",
    properties = {
        username = { type = "string" },
        password = { type = "string", encrypted = true },   <==== Here
    },
    required = {"username", "password"},    
}
```

3. Set the plugin

```Shell
curl http://127.0.0.1:9180/apisix/admin/consumers -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "username": "foo",
    "plugins": {
        "basic-auth": {
            "username": "foo",
            "password": "bar"       <= Sensitive Information
        }
    }
}'
```

4. The plugin data read directly from ETCD is encrypted

```C#
etcdctl get /apisix/consumers/foo
{
    "username":"foo",
    "update_time":1668584767,
    "create_time":1668584767,
    "plugins":{
        "basic-auth":{
            "username":"foo",
            "password":"da61c0083b6d19ef3db2490d0da96a71572da0fa"  <= encrypted
        }
    }
}
```

5. The data surface is read normally

```Bash
curl http://127.0.0.1:9180/apisix/admin/consumers -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1'
{
  "list": [
    {
      "createdIndex": 8627,
      "modifiedIndex": 8627,
      "value": {
        "username": "foo",
        "update_time": 1668584767,
        "create_time": 1668584767,
        "plugins": {
          "basic-auth": {
            "username": "foo",
            "password": "bar"         <= Unencrypted
          }
        }
      },
      "key": "/apisix/consumers/foo"
    }
  ],
  "total": 1
}
```

# Detailed Design

1. Add a field in the configuration file to enable data encryption

```YAML
apisix:
    data_encryption:                # use `encrypted = {fields}` in plugin schema to enable encryption
        enable: false               # If not set, the default value is `false`.
        keyring:
            - edd1c9f0985e76a2      # If not set, will save origin value into etcd.
            - qeddd145sfvddff3      # If set this, the keyring should be an array whose elements are string, and the size is also 16, and it will encrypt fields with AES-128-CBC
                                    # !!! So do not change it after encryption, it can't decrypt the fields have be saved if you change !!
                                    # Only use the first key to encrypt, and decrypt in the order of the array.
```

2. Specify the fields to be encrypted in the schema of the plugin

```Ada
local consumer_schema = {
    type = "object",
    title = "work with consumer object",
    properties = {
        username = { type = "string" },
        password = { type = "string", encrypted = true },    <===== here
    },
    required = {"username", "password"},
}

local schema = {

}
```

3. If data encryption is enabled, the corresponding fields of the following plugins will be encrypted

```Bash
jwt-auth：secret，private_key
basic-auth：password
authz-keycloak：client_secret 
authz-casdoor：client_secret
openid-connect：client_secret
hmac-auth: secret_key
csrf：key
rocketmq-logger：secret_key
clickhouse-logger：password
error-log-logger：clickhouse.password
sls-logger：access_key_secret
google-cloud-logging：auth_config.private_key
elasticsearch-logger：auth.password
tencent-cloud-cls：secret_key
kafka-proxy：sasl.password
```

4. Encryption and Decryption

- Commonly used algorithms AES, ChaCha20, we use AES128CBC encryption algorithm, the logic is the same as ssl encryption and decryption, that is, only use the first key to encrypt, if it fails to take turns to use all keys to decrypt, consider reusing the ssl encryption and decryption function  （https://github.com/apache/apisix/blob/master/apisix/ssl.lua）

- Currently, only those distributed through the apisix admin interface can be encrypted. Many users have developed their own admin and directly manipulated storage media such as etcd, so they need to implement the encryption function themselves.

- After the encryption function is enabled, the data that was not encrypted needs to be sent down again through the interface in order to be encrypted, is a refresh interface provided? Consider not providing it for now

- Combining the above, the original string is used directly after decryption failure

- Admin now only uses etcd, considering that the first version only supports etcd and only supports encryption of plugin parameters

- Encryption：

Add encryption and decryption to check_single_plugin_schema, note that there are 2 different types of schema within the plugin, namely: schema and consumer_schema

```Lua
local function check_single_plugin_schema(name, plugin_conf, schema_type, skip_disabled_plugin, encrypt_or_decrypt)
    ...
    if plugin_obj.check_schema then
        ...
    end

    if schema_type == core.schema.TYPE_CONSUMER then
        ok, err = encrypt(plugin_conf, plugin_obj.consumer_schema, encrypt_or_decrypt)
    else
        ok, err = encrypt(plugin_conf, plugin_obj.schema, encrypt_or_decrypt)
    end

    return true
end
```

The encryption and decryption function iterates through the corresponding schema, and if it finds the encrypted = true field, it encrypts and decrypts the field in the corresponding configuration

```Lua
local function encrypt(plugin_conf, schema, encrypt_or_decrypt)
    if schema.type == "object" then
        for name, obj in pairs(schema.properties) do
            encrypt(plugin_conf[name], obj, encrypt_or_decrypt)
        end
    elseif schema.type = "array" then
        ...
    end
    
    if schema.encrypted then
        aes128(plugin_conf)
    end
end
```

Add corresponding options in check_schema and stream_check_schema, now directly called in admin, no need to modify, encrypt_or_decrypt is nil by default, i.e. encrypt

```Lua
local function check_schema(plugins_conf, schema_type, skip_disabled_plugin, encrypt_or_decrypt)
    for name, plugin_conf in pairs(plugins_conf) do
        local ok, err = check_single_plugin_schema(name, plugin_conf,
            schema_type, skip_disabled_plugin, encrypt_or_decrypt)
        if not ok then
            return false, err
        end
    end

    return true
end
```

plugin_checker and stream_plugin_checker are used for checksumming when reading data from etcd, and are decrypted by default

```Ada
function _M.plugin_checker(item, schema_type)
    if item.plugins then
        return check_schema(item.plugins, schema_type, true, decrypt) <=== decrypt
    end

    return true
end
```

### PR DESCRIPTION
**Pull request by tzssangglass**
### Description

Fixes #8407

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by soulbird** [comment]
It is best to update the documentation of each plugin to indicate the fields that will be encrypted.

**Comment by tzssangglass** [comment]
> It is best to update the documentation of each plugin to indicate the fields that will be encrypted.

done

File: apisix/plugin.lua [hunk 1]
```diff
@@ -903,19 +926,57 @@ _M.decrypt_conf = decrypt_conf
...
     if not schema then
+        core.log.warn("failed to get schema for plugin: ", name)
         return
     end
 
-    for key, props in pairs(schema.properties) do
-        if props.type == "string" and props.encrypted and conf[key] then
-            local encrypted = apisix_ssl.aes_encrypt_pkey(conf[key], "data_encrypt")
-            conf[key] = encrypted
+    if schema.encrypt_fields and core.table.nkeys(schema.encrypt_fields) > 0 then
```
Comment (spacewander) [inline apisix/plugin.lua:933]: Better to use `core.table.isempty`?

File: docs/en/latest/plugin-develop.md [hunk 2]
```diff
@@ -300,13 +300,21 @@ Specify the parameters to be stored encrypted. (Requires APISIX version >= 3.1.0
...
-Parameters can be stored encrypted by specifying `encrypted = true` in the `schema`. APISIX will provide the following functionality.
+If it is a nested parameter, such as the `clickhouse.password` parameter of the `error-log-logger` plugin, it needs to be separated by `.`:
 
-- When adding and updating resources via the `Admin API`, APISIX automatically encrypts parameters with `encrypted = true` and stores them in etcd
-- When fetching resources via the `Admin API` and when running the plugin, APISIX automatically decrypts the `encrypted = true` parameter
+```lua
+encrypt_fields = {"clickhouse.password"}
+```
+
+Currently only two levels of nesting are supported.
```
Comment (spacewander) [inline docs/en/latest/plugin-develop.md:312]: Let's make it clear that we don't support fields in array

File: docs/zh/latest/plugins/openid-connect.md [hunk 3]
```diff
@@ -61,6 +61,8 @@ description: OpenID Connect（OIDC）是基于 OAuth 2.0 的身份认证协议
 | session                              | object  | 否     |                       |               | 当设置 bearer_only 为 false 时，openid-connect 插件将使用 Authorization Code 在 IDP 上进行认证，因此你必须设置 session 相关设置。 |
 | session.secret                       | string  | 是     | 自动生成               | 16 个以上字符  | 用于 session 加密和 HMAC 计算的密钥。 |
 
+NOTE: `encrypt_fields = {"client_secret"}` is also defined in the schema, which means that the field will be stored encrypted in etcd. See [encrypted storage fields](../plugin-develop.md#encrypted-storage-fields).
```
Comment (spacewander) [inline docs/zh/latest/plugins/openid-connect.md:64]: Let's use Chinese

File: docs/en/latest/plugins/openid-connect.md [hunk 4]
```diff
@@ -61,6 +61,8 @@ description: OpenID Connect allows the client to obtain user information from th
 | session                              | object  | False    |                       |              | When bearer_only is set to false, openid-connect will use Authorization Code flow to authenticate on the IDP, so you need to set the session-related configuration. |
 | session.secret                       | string  | True     | Automatic generation  | 16 or more characters | The key used for session encrypt and HMAC operation. |
 
+注意：schema 中还定义了 `encrypt_fields = {"client_secret"}`，这意味着该字段将会被加密存储在 etcd 中。具体参考 [加密存储字段](../plugin-develop.md#加密存储字段)。
```
Comment (spacewander) [inline docs/en/latest/plugins/openid-connect.md:64]: Let's use English

File: apisix/plugin.lua (hunk 1 above)
Comment (tzssangglass) [reply apisix/plugin.lua]: done

File: docs/zh/latest/plugins/openid-connect.md (hunk 3 above)
Comment (tzssangglass) [reply docs/zh/latest/plugins/openid-connect.md]: done

File: apisix/plugin.lua [hunk 5]
```diff
@@ -903,19 +926,57 @@ _M.decrypt_conf = decrypt_conf
...
+                end
+            elseif core.string.find(key, ".") then
+                -- encrypt fields has indents
+                local res, err = re_split(key, "\\.", "jo")
+                if not res then
+                    core.log.warn("failed to split key [", key, "], err: ", err)
+                    return
+                end
+
+                -- we only support two levels
```
Comment (monkeyDluffy6017) [inline apisix/plugin.lua:951]: So many limitations, is it written in doc?

Comment (monkeyDluffy6017) [reply apisix/plugin.lua:951]: Why add this limitation, can we parse the conf recursively

File: docs/en/latest/plugins/csrf.md [hunk 6]
```diff
@@ -41,6 +41,8 @@ This Plugin considers the `GET`, `HEAD` and `OPTIONS` methods to be safe operati
 | expires | number | False    | `7200`              | Expiration time in seconds of the CSRF cookie. Set to `0` to skip checking expiration time. |
 | key     | string | True     |                     | Secret key used to encrypt the cookie.                                                      |
 
+NOTE: `encrypt_fields = {"key"}` is also defined in the schema, which means that the field will be stored encrypted in etcd. See [encrypted storage fields](../plugin-develop.md#encrypted-storage-fields).
```
Comment (monkeyDluffy6017) [inline docs/en/latest/plugins/csrf.md:44]: The field will be encrypted and stored in etcd ?

Comment (tzssangglass) [reply docs/en/latest/plugins/csrf.md:44]: yes, what is stored in etcd is the encrypted

Comment (monkeyDluffy6017) [reply docs/en/latest/plugins/csrf.md:44]: Is the sentence better `The field will be encrypted and stored in etcd `

Comment (tzssangglass) [reply docs/en/latest/plugins/csrf.md:44]: I guess there is no difference? For this field, the only encrypted state is in etcd. They are decrypted when APISIX is running and when fetched via the admin API.
I think the way I wrote it better conveys the idea that it is only encrypted in etcd.

**Review by soulbird — APPROVED** [review]

File: apisix/plugin.lua [hunk 7]
```diff
@@ -903,19 +926,57 @@ _M.decrypt_conf = decrypt_conf
...
         end
     end
 end
 _M.encrypt_conf = encrypt_conf
 
 
+check_plugin_metadata = function(item)
+    local ok, err = check_single_plugin_schema(item.id, item,
+                                               core.schema.TYPE_METADATA, true)
+    if ok and enable_gde()then
```
[the remark this answers is not in the document]
Comment (tzssangglass) [reply apisix/plugin.lua]: done

File: docs/en/latest/plugins/key-auth.md [hunk 8]
```diff
@@ -41,8 +41,6 @@ For Consumer:
 |------|--------|-------------|----------------------------|
 | key  | string | required    | Unique key for a Consumer. |
 
-NOTE: The schema for `key` also defines `encrypted = true`, which means that the field will be stored encrypted in etcd. See [encrypted storage fields](../plugin-develop.md#encrypted-storage-fields).
```
[the remark this answers is not in the document]
Comment (tzssangglass) [reply docs/en/latest/plugins/key-auth.md:44]: type, fix it

File: docs/zh/latest/plugins/key-auth.md [hunk 9]
```diff
@@ -41,8 +41,6 @@ Consumer 端：
 | ---- | ------ | ------ | ------------------------------------------------------------------------------------------------------------- |
 | key  | string | 是     | 不同的 Consumer 应有不同的 `key`，它应当是唯一的。如果多个 Consumer 使用了相同的 `key`，将会出现请求匹配异常。 |
 
-注意：`key` 的 schema 中还定义了 `encrypted = true`，这意味着该字段将会被加密存储在 etcd 中。具体参考 [加密存储字段](../plugin-develop.md#加密存储字段)。
```
[the remark this answers is not in the document]
Comment (tzssangglass) [reply docs/zh/latest/plugins/key-auth.md:44]: typo, fix it

File: docs/zh/latest/plugins/key-auth.md [hunk 10]
```diff
@@ -41,6 +41,8 @@ Consumer 端：
 | ---- | ------ | ------ | ------------------------------------------------------------------------------------------------------------- |
 | key  | string | 是     | 不同的 Consumer 应有不同的 `key`，它应当是唯一的。如果多个 Consumer 使用了相同的 `key`，将会出现请求匹配异常。 |
 
+注意：schema 中还定义了 `encrypt_fields = {"client_secret"}`，这意味着该字段将会被加密存储在 etcd 中。具体参考 [加密存储字段](../plugin-develop.md#加密存储字段)。
```
Comment (spacewander) [inline docs/zh/latest/plugins/key-auth.md:44]: Bad field name?

Comment (tzssangglass) [reply docs/zh/latest/plugins/key-auth.md]: fix

**Review by soulbird — APPROVED** [review]

**Review by monkeyDluffy6017 — APPROVED** [review]

**Review by spacewander — APPROVED** [review]

**Pull request merged**